### DISCUSSION TITLE
RavenDB restarts (crashes?) under load

### DISCUSSION THREAD

**Discussion by Shaddix** [ravendb/ravendb#18749] [Q&A]
I'm running Community edition of RavenDB 5.4.200 in a single-node mode on AWS EC2 (m6a.xlarge) via Fargate task (i.e. in Docker).
I have a fanout map-reduce index (which outputs results to a Collection), which produces 100 entries from a single document.

Everything is working fine under low load.
Though, during a stress-test, when I create 8k documents (where each fans out 100 items) in 4 parallel processes, RavenDB restarts at about 3k of such docs.

After restart it continues the indexing and usually successfully indexes all documents, but I still want to avoid the restart.

When watching the state of the index at the time before restart, I could notice that the number of entries stops incrementing for prolonged period of time), which might indicate that something goes wrong.
![image](https://github.com/ravendb/ravendb/assets/1249513/886f5e80-f857-4195-b21b-970ae6c2a119)

Can I tweak the settings somehow to avoid restarts?
Is RavenDB team interested in supporting a user of Community-edition database?

I could reproduce this restart (crash?) easily, and if needed I could provide any logs or even access to the EC2 instance.
Though, logs at the information level don't provide much info (at least for me), here's the portion at the time of restart:
![image](https://github.com/ravendb/ravendb/assets/1249513/25fe33e9-3fa3-40f3-b700-c572a7121aa1)

**Comment by ayende** [comment]
What does you index looks like?
Can you check the system log?

**Reply by Shaddix** [reply]
Here's the index:
```
map('StockCountEventBatchV6s', (batch) => {
  return batch.Events.map(e => {
    let parentThingId = e.ParentThingId;
    let expectedQuantity = 0;
    let externalReference = e.ExternalReference;
    let stockCountId = e.StockCountId;
    let splitAttribute = e.SplitAttribute;
    let stockCountSessionId = e.StockCountSessionId;

    return {
        ThingId: e.ThingId,
        ParentThingId: parentThingId,
        IsMatchScopeAssetFilter: e.IsMatchScopeAssetFilter,
        IsMatchGlobalAssetFilter: e.IsMatchGlobalAssetFilter,
        StockCountId: stockCountId,
        StockCountSessions: [stockCountSessionId],
        StockCountSessionId: stockCountSessionId,
        ZoneId: e.ZoneId,
        SubZoneId: e.SubZoneId,
        Zone: e.Zone,
        SubZone: e.SubZone,
        ServerTimestamp: e.ServerTimestamp,
        TeamId: e.TeamId,
        ParentAttributes: e.Attributes,
        //GlobalExpected: globalExpected,
        ExternalReference: externalReference,
        SplitAttributeName: splitAttribute,
        Duplicates: 1,
        ScopedExpectedQuantitiesId: e.ScopedExpectedQuantitiesId,
        StockCountScopeId: e.StockCountScopeId,
    }
  });
})

groupBy(e => ({
    ThingId: e.ThingId,
    StockCountId: e.StockCountId,
})).aggregate(g => {
    let e = g.values[0];
    var firstSeenAt = g.values.reduce((result, item) => (!result.ServerTimestamp || item.ServerTimestamp < result.ServerTimestamp) ? item : result, {});
    return {
        ThingId: g.key.ThingId,
        ParentThingId: e.ParentThingId,
        IsMatchScopeAssetFilter: e.IsMatchScopeAssetFilter,
        IsMatchGlobalAssetFilter: e.IsMatchGlobalAssetFilter,
        StockCountId: e.StockCountId,
        StockCountSessionId: firstSeenAt.StockCountSessionId,
        ServerTimestamp: firstSeenAt.ServerTimestamp,
        StockCountSessions: [...new Set([].concat(...g.values.map(x=>x.StockCountSessions)))],
        TeamId: e.TeamId,
        ParentAttributes: e.ParentAttributes,
        SubZoneId: e.SubZoneId,
        ZoneId: e.ZoneId,
        Zone: e.Zone,
        SubZone: e.SubZone,
        //GlobalExpected: globalExpected,
        ExternalReference: e.ExternalReference,
        SplitAttributeName: e.SplitAttributeName,
        Duplicates: g.values.reduce((result, item) => result + item.Duplicates, 0),
        ScopedExpectedQuantitiesId: e.ScopedExpectedQuantitiesId,
        StockCountScopeId: e.StockCountScopeId,
    };
})
```

**Reply by Shaddix** [reply]
[messages.txt](https://github.com/user-attachments/files/16024366/messages.txt)
And here's the syslog.
Indeed, there's out of memory in the logs:

> Jun 27 20:08:43 ip-10-0-152-199 kernel: Memory cgroup out of memory: Killed process 1337 (Raven.Server) total-vm:367216416kB, anon-rss:12265576kB, file-rss:1696556kB, shmem-rss:78196kB, UID:0 pgtables:147772kB oom_score_adj:0

**Reply by Shaddix** [reply]
maybe I misconfigured something, but the EC2 itself has 16GB RAM, and assuming the license allows 6GB I didn't expect memory issues on the OS level

**Reply by ayende** [reply]
I think that the issue is your container configuration, see: https://stackoverflow.com/a/67261356/6366

Basically, you should exempt RavenDB from the OOM killer. RavenDB will try to use as much memory as you give it, but if there is a memory pressure, the process with the higher % of memory utilized will be killed first.

**Reply by ayende** [reply]
Note that `anon-rss:12265576kB` _is_ kind of high. 

Looking at the reduce:

```
        StockCountSessions: [...new Set([].concat(...g.values.map(x=>x.StockCountSessions)))],
```

This looks suspicious. I assume that you intend to have the unique `StockCountSessions` ? What is that, and id?
Can you check that you get the right values there?

When you look at the memory, do you see high values in the unmanaged area?
Can you send the debug package?

**Reply by Shaddix** [reply]
> This looks suspicious. I assume that you intend to have the unique StockCountSessions ? What is that, and id?
> Can you check that you get the right values there?
I have several tests around that and it's correct.

The idea that the same ThingId could be present in different Sessions (and even multiple times within the same Session), and I want to store distinct Session Ids (StockCountSessions is an array of Id's, yes)

**Reply by ayende** [reply]
How many of those may you have?

**Reply by Shaddix** [reply]
> I think that the issue is your container configuration, see: https://stackoverflow.com/a/67261356/6366
> 
> Basically, you should exempt RavenDB from the OOM killer. RavenDB will try to use as much memory as you give it, but if there is a memory pressure, the process with the higher % of memory utilized will be killed first.

thanks, I'll try that! Though, EC2 has the single running container (the RavenDB) and there's nothing else running on this EC2, so nothing else should consume memory

>Can you send the debug package?
Do you mean the sourcecode / compiled binaries? It's too tied to AWS services, it'd be really hard to run them outside of our infrastructure.

**Reply by Shaddix** [reply]
>How many of those may you have?
usually not much (below 100 under the same StockCountId), in this stress-test there's actually 4 StockCounts with a single session in each.

**Reply by Shaddix** [reply]
>When you look at the memory, do you see high values in the unmanaged area?

The memory state was like that when everything was running normally (around 3GB, sometimes dropping to 500MB and returning back to 3GB)
![image](https://github.com/ravendb/ravendb/assets/1249513/7c8255a6-df1b-48e4-ba7f-dc056ce7d3c2)

Just before the OOM the Unmanaged Allocations started to increase and within 1 minute reached the state like that (and then it was killed):
![image](https://github.com/ravendb/ravendb/assets/1249513/3d88e4b7-8bf3-4045-a73b-760ac4e15be1)

**Reply by Shaddix** [reply]
Seems like [this discussion](https://github.com/aws/containers-roadmap/issues/753) shows it's problematic to set something like `--oom-kill-disable` for Fargate tasks running on EC2..

Am I right, that there'd be no issues if RavenDB is running on EC2 itself without Docker/Fargate?

**Reply by ayende** [reply]
Yes, that looks like to be the root cause.
We are trying to consume the memory we have avialable, but we cannot access that.
EC2 would be fine, since it won't compete with other things.

That said, `Unamanged Allocations` is pretty high .
Can you look at the debug package when this happens, should tell us where the cost is.

**Reply by Shaddix** [reply]
~1 minute before raven get's killed
![image](https://github.com/ravendb/ravendb/assets/1249513/8ea5cb80-e9ed-48fb-be46-26f6b83ca7df)
[2024-06-29 7-14-03 - Node [A].zip](https://github.com/user-attachments/files/16039157/2024-06-29.7-14-03.-.Node.A.zip)

This is 2 minutes before that when it seems to operate normally, with Unmanaged allocations around 3GB
[2024-06-29 7-12-00 - Node [A].zip](https://github.com/user-attachments/files/16039170/2024-06-29.7-12-00.-.Node.A.zip)

I'm gonna install RavenDB on EC2 directly and check if it solves the issue

Thanks a lot for help!

**Reply by ayende** [reply]
This index is *quite* expensive: `Index_StockCountEventsEnrichedV6`

I'm seeing:

```
                   "InputCount": 1,
                    "FailedCount": 0,
                    "OutputCount": 200,
                    "SuccessCount": 1,
                    "Started": "2024-06-29T07:06:04.7159010Z",
                    "DurationInMs": 1191.23,
                    "AllocatedBytes": {
                        "SizeInBytes": 19553768,
                        "HumaneSize": "18.65 MBytes"
                    },
```

So a single document caused us to do map/reduce on 200 items.

In particular, you have this line in the Map:

```
  return batch.Events.map(e => {
```

Which means that you are doing a fanout. And a single `StockCountEventBatchV6` will generate a lot of pressure during indexing.

I assume that this is intentional, but that makes more sense in terms of the memory usage.

So the next stage is to ensure that the OS won't kill us needlessly.

**Reply by Shaddix** [reply]
the fanout was intentional, since the indexing was faster this way (comparing to storing each event in a separate document).

I tried running it without docker on EC2 and it was alive longer, but raven was still killed at some point with `.NET TP Worker invoked oom-killer` (I used
`echo 1 > /proc/sys/vm/overcommit_memory`, though, it was killed with 0 and 2 in `overcommit_memory` as well).
[messages.txt](https://github.com/user-attachments/files/16044160/messages.txt)

Do I miss some other OS-level settings?..

I will also try reducing the batch size to check at what value it'd be ok

**Reply by ayende** [reply]
How much memory do you have total on the machine?
Is it possible that you hit some _really_ full batch? In the example above, you have 1 -> 200 rate of change.
If you have 100 documents being changed, that would be 20,000 items that we have to map & reduce.

**Reply by Shaddix** [reply]
EC2 has 16GB ram (m6a.xlarge)
The maximum size of a batch is 200, batches never change, but there are a lot of them being sent in parallel, that's true (i.e. in ~15 minutes about 400 of such batches are created)

**Reply by ayende** [reply]
That doesn't sound right, but I believe this is mostly about your index creating a lot of memory pressure as it works.
A rate of 1:200 can certainly cause us to try an operation that would kill the memory.
you can try setting the `MaxMapCount` setting, to a lower value. To reduce the size of batches.

**Reply by Shaddix** [reply]
Do you mean `Indexing.MapBatchSize`? What value could you recommend? Is e.g. 1024 ok?

**Reply by ayende** [reply]
Yes, that is the one.
Before recommending anything, can you look at the index performance?

![image](https://github.com/ravendb/ravendb/assets/116915/f9b2b56c-ced4-4616-8ef5-ae3248966a08)

Send us the export, that would tell us a lot about what is actually going on and allow to correlate sizes to memory.
